distinct 'static' items never overlap (#1657)
rust-lang/reference: src/items/static-items.md

> [^extern-safety]: The `safe` and `unsafe` function qualifiers are only
>   allowed semantically within `extern` blocks.

A *static item* is similar to a [constant], except that it represents an allocated object in the
program that is initialized with the initializer expression. This allocated object is disjoint from

> all other allocated objects. All references and raw pointers to the static refer to the same

Reviewer: Do we want to guarantee that they are disjoint from other allocated objects? Or just other statics. IE. if I have:
```rust
static FOO: i32 = 0;
const BAR: i32 = 0;

fn foo(){
    assert!(!core::ptr::eq(&FOO, &BAR));
}
```

Should we guarantee the assertion will always pass?

Reviewer: I don't have concern permissions, but I think this should be addresses by T-opsem before the end of the FCP.

Author: Yeah I think we should guarantee that, and it's what the text already says, isn't it?

Reviewer 2: That prevents emitting `unnamed_addr` on `const` items.

Reviewer 2: `unnamed_addr` specifically *does* coalescing of non-significant-address-items into each other, and potentially into a significant-address item.

Author: Wait, LLVM will merge `unnamed_addr` allocations with "named addr" (non-`unnamed_addr`) allocations? That's... surprising.

Cc @nikic

Reviewer 3: The LangRef says:
> Note that a constant with significant address can be merged with a unnamed_addr constant, the result being a constant whose address is significant.

Author: Ah, that is unfortunate.

Reviewer: > Yeah I think we should guarantee that, and it's what the text already says, isn't it?

That's what the proposed text says, my question is whether that's what we want it to say.

Reviewer 2: I... honestly don't think that's surprising at all?

That's exactly what I would expect in a case of
```rust
const BIG_CONST: BigFrozen = BigFrozen::big_init();
static BIG_STATIC: BigFrozen = BIG_CONST;
```

That's the precise situation where the `const` will be unified "into" the `static`.

Author: > I... honestly don't think that's surprising at all?

I guess we have different intuitions then.

But reality clearly works one way here, so I have updated the text to match reality. @chorman0773 please have a look.

> [^extern-safety]: The `safe` and `unsafe` function qualifiers are only
>   allowed semantically within `extern` blocks.

A *static item* is similar to a [constant], except that it represents an allocated object in the
program that is initialized with the initializer expression. All references and raw pointers to the
static refer to the same allocated object. Static items have the `static` lifetime, which outlives
all other lifetimes in a Rust program. Static items do not call [`drop`] at the end of the program.

This allocated object is disjoint from all other `static` objects as well as heap-allocated and
stack-allocated variables. However, the storage of immutable `static` items can overlap with objects

> that do not themselves have a unique address, such as promoteds and `const` items.

Reviewer 2: this seems basically good to me, thank you!

though I wonder if https://github.com/rust-lang/reference/blob/23ce619966541bf2c80d45fdfeecf3393e360a13/src/destructors.md?plain=1#L295 should be linked from [promoteds]?

>   allowed semantically within `extern` blocks.

r[items.static.intro]
A *static item* is similar to a [constant], except that it represents an allocated object in the
program that is initialized with the initializer expression. All references and raw pointers to the
static refer to the same allocated object.

r[items.static.lifetime]
Static items have the `static` lifetime, which outlives all other lifetimes in a Rust program.
Static items do not call [`drop`] at the end of the program.

r[items.static.storage-disjointness]
This allocated object is disjoint from all other `static` objects as well as heap-allocated and
stack-allocated variables. However, the storage of immutable `static` items can overlap with objects
that do not themselves have a unique address, such as [promoteds] and [`const` items][constant].

r[items.static.namespace]
The static declaration defines a static value in the [value namespace] of the module or block where it is located.

r[items.static.init]
The static initializer is a [constant expression] evaluated at compile time.

> Static initializers may refer to and even read from other statics.

Reviewer 4: ```suggestion
Static initializers may refer to and read from other statics.
```

This feels out of place.

>   allowed semantically within `extern` blocks.

r[items.static.intro]
A *static item* is similar to a [constant], except that it represents an allocated object in the
program that is initialized with the initializer expression. All references and raw pointers to the
static refer to the same allocated object.

r[items.static.lifetime]
Static items have the `static` lifetime, which outlives all other lifetimes in a Rust program.
Static items do not call [`drop`] at the end of the program.

r[items.static.storage-disjointness]

> If the `static` has a size of at least 1 byte, this allocated object is disjoint from all other

Reviewer 5: ```suggestion
If the `static` has a size of at least 1 byte, this allocated object is disjoint from all other such
```

To make it clear that this applies to other statics with a size of at least one byte.

> `static` objects as well as heap-allocated and stack-allocated variables. However, the storage of

Reviewer 6: What’s a “heap-allocated variable”? Presumably, things like the contents of `Box`es, but calling that a variable seems too informal and imprecise to be appropriate for the reference — especially as the reference already [defines a variable as](https://doc.rust-lang.org/reference/variables.html) “a component of a stack frame”.

Perhaps something like:

```suggestion
`static` objects as well as heap allocations and stack allocations. However, the storage of
```

> immutable `static` items can overlap with objects that do not themselves have a unique address, such

Reviewer 6: “Immutable `static` item” could be interpreted as meaning a static item without the `mut` keyword, but it must also exclude interior-mutable memory. Perhaps this should be clarified (or perhaps it doesn’t matter since this is a “can”, not “must”).

Author: Yes we should have this as a shared definition somewhere that we can reference.
